src/librepgp/stream-parse.cpp

```diff
@@ -797,40 +810,56 @@ signed_src_update(pgp_source_t *src, const void *buf, size_t len)
     if (param->txt_hashes.empty()) {
         return;
     }
-    /* check whether we had CR at the end of last chunk and LF at the beginning */
+
     uint8_t *ch = (uint8_t *) buf;
-    if (param->lastcr && (*ch == CH_LF)) {
-        ch++;
-    }
     uint8_t *linebeg = ch;
     uint8_t *end = (uint8_t *) buf + len;
-    /* we support CR, LF and CRLF line endings */
+    bool     max_linelen_warn = false;
+    /* we support LF and CRLF line endings */
     while (ch < end) {
-        /* continue if not reached CR or LF */
-        if ((*ch != CH_CR) && (*ch != CH_LF)) {
+        /* continue if not reached LF */
+        if (*ch != CH_LF) {
+            if (*ch != CH_CR && param->stripped_crs > 0) {
+                while (param->stripped_crs--)
+                    pgp_hash_list_update(param->txt_hashes, ST_CR, 1);
+
+                param->stripped_crs = 0;
+            }
+
+            if (!max_linelen_warn && (ch - linebeg) >= MAXIMUM_GNUPG_LINELEN) {
```

Comment:
Do we assume here that the line starts since `buf` or later?
What would happen if a line started in a previous `signed_src_update` call?

Reply:
Yes, right, data is processed in 32k blocks, not line by line, and line started in previous block could bypass this check. Going to fix that.

Reply:
...and add some test for this exact case (and probably 8k-16k-64k blocks) so we'll not regress in future )

Reply:
@ni4 how exactly this test should look like? cli test that checks if warning message is printed on stderr?

Reply:
@antonsviridenko It should just process test data file which exposes the mentioned behaviour (i.e. line started on a previous `signed_src_update()`), if this is possible.

Reply:
Just to be clear, I'm talking about this piece (updated) only. It appears that github's web ui shows less lines mentioned than in email notifications

```
            if (!param->max_line_warn && param->text_line_len >= MAXIMUM_GNUPG_LINELEN) {
                RNP_LOG("Canonical text document signature: line is too long, may cause "
                        "incompatibility with other implementations. Consider using binary "
                        "signature instead.");
                param->max_line_warn = true;
            }
```
Otherwise the code already handles text lines that cross 32k block boundary, and it is covered by existing tests
https://github.com/rnpgp/rnp/blob/master/src/tests/ffi.cpp#L7028
and "cli_tests.SignDefault" where random text document up to 1M is generated each time and therefore with high probability many lines cross the block boundary.

Reply:
And rnp CLI does not support detached text mode signatures yet, so it will be bit problematic to generate one :) If we sign that test file with text lines longer than 19995 characters using GnuPG, it will be verified as BAD signature by rnp. Because GnuPG forcefully inserts CRLF after 19995 chars.